File: js/src/maplestory/api/common/dto/character/characterSetEffect.ts
export abstract class CharacterSetEffectDto {
  public abstract date: Date | null;
  public abstract setEffect: CharacterSetEffectSetDto[];
}

export abstract class CharacterSetEffectInfoDto {
  public abstract setCount: number;
  public abstract setOption: string;
}

export abstract class CharacterSetEffectOptionFullDto {
  public abstract setCount: number;
  public abstract setOption: string;
}

export abstract class CharacterSetEffectSetDto {
  public abstract setName: string;
  public abstract totalSetCount: number;
  public abstract setEffectInfo: CharacterSetEffectInfoDto[];
  public abstract setOptionFull: CharacterSetEffectOptionFullDto[];
}

File: csharp/MapleStory.OpenAPI/Src/Common/DTO/Character/CharacterSetEffect.cs
namespace MapleStory.OpenAPI.Common.DTO
{
    public abstract class CharacterSetEffectDTO<TCharacterSetEffectSet>
    {
        public abstract DateTimeOffset? Date { get; set; }
        public abstract List<TCharacterSetEffectSet> SetEffect { get; set; }
    }

    public abstract class CharacterSetEffectDTO : CharacterSetEffectDTO<CharacterSetEffectSetDTO>
    {

    }

    public abstract class CharacterSetEffectInfoDTO
    {
        public abstract int SetCount { get; set; }
        public abstract string SetOption { get; set; }
    }

    public abstract class CharacterSetEffectOptionFullDTO
    {
        public abstract int SetCount { get; set; }
        public abstract string SetOption { get; set; }
    }

    public abstract class CharacterSetEffectSetDTO<TCharacterSetEffectInfo, TCharacterSetEffectOptionFull>
    {
        public abstract string SetName { get; set; }
        public abstract long TotalSetCount { get; set; }
        public abstract List<TCharacterSetEffectInfo> SetEffectInfo { get; set; }
        public abstract List<TCharacterSetEffectOptionFull> SetOptionFull { get; set; }
    }

    public abstract class CharacterSetEffectSetDTO : CharacterSetEffectSetDTO<CharacterSetEffectInfoDTO, CharacterSetEffectOptionFullDTO>
    {

    }
}
